### PR TITLE
Bump kubernetes to 1.28.4

### PR DESCRIPTION
**Pull request by farodin91**

### CODE REVIEW

**Review by majolo — APPROVED** [review]
This looks good to me, would be great to get a maintainer to approve and merge this if they're happy :+1:

**Comment by crenshaw-dev** [comment]
Thanks so much for picking this up, @majolo! I'm going to wait until we cut Argo CD 2.10 before tackling this, since we have some other high-priority changes we want to get in.

**Comment by crenshaw-dev** [comment]
@farodin91 the earliest will be > next Monday.

I'll go ahead and give a quick first-pass review so we can knock out any basic concerns.

File: pkg/utils/kube/kube.go
Comment (crenshaw-dev) [inline pkg/utils/kube/kube.go:409]: Just curious, why `PollUntilContextCancel` here instead of `PollUntilContextTimeout?`

File: pkg/utils/kube/resource_ops.go
Comment (crenshaw-dev) [inline pkg/utils/kube/resource_ops.go:271]: Do you know why these fields were removed from the upstream struct? Do we need to be switching to some other mechanism of providing these arguments?

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: pkg/utils/kube/resource_ops.go
Comment (majolo) [reply pkg/utils/kube/resource_ops.go:271]: I made a similar change in our fork - without much digging, the following seemed to suggest just removing these is the correct approach?

https://github.com/kubernetes/kubernetes/pull/114294/files
https://github.com/kubernetes/kubernetes/pull/114413/files

**Comment by majolo** [comment]
Thanks @crenshaw-dev could this not be cut first and then included in argo-cd 2.10? It's actually argo-cd that we use (which in turn imports this library) which is causing us the issues.

**Comment by crenshaw-dev** [comment]
@majolo theoretically, yes. But practically speaking, no. Before this can be merged, a similar PR will need to be made on the argo-cd repo and reviewed/tested. With those two PRs together, I just don't think I'll have time to review/merge before we cut RC1 Monday.

File: pkg/utils/kube/kube.go
Comment (farodin91) [reply pkg/utils/kube/kube.go:409]: I would need a timeout and i didn't had one before.

**Comment by dfroberg** [comment]
Any chance of getting this one bumped together with this one:
https://github.com/argoproj/gitops-engine/issues/558

**Comment by majolo** [comment]
@crenshaw-dev any update on this one? Thanks!

**Comment by asaf-erlich** [comment]
Hello 👋 , our team owns an internal tool that interacts with both argo cd and k8s. As part of the recent upgrade a different team is making to all our k8s clusters to 1.27+ I'm trying to upgrade our tool's internal libraries. When doing so the gitops-engine dependency, even if I bump it to the sha in main/master is not able to compile and is blocking the upgrade. 

This PR's code works perfectly with no compile errors (I checked out the forked code locally). I would love to know a rough timeline of when this could get merged? Alternatively if you're willing to push a tag or something so my go.mod file could depend on it. This would be greatly appreciated 🙏

**Comment by crenshaw-dev** [comment]
@asaf-erlich we can start working on getting this merged now.

Would you mind rebasing?

Could you also open a PR on argo-cd pointing to the latest commit of your fork/branch? Here's an example of what that change looks like. https://github.com/argoproj/argo-cd/pull/16858/commits/90ce09a75ee7136d08bc11d9c89018e2353d6652

That PR will let us run Argo CD's tests and catch any issues which weren't revealed by the gitops-engine tests.

**Comment by crenshaw-dev** [comment]
Looks like we could actually go to 1.28.6 or to 1.29.1 at this point.

**Comment by asaf-erlich** [comment]
@crenshaw-dev I'm not the original author of the PR, @farodin91 is. I'm happy to attempt those things it will help but would need to create a new PR to do so. I'd like to give @farodin91 at least a day or two to respond and if they cannot do it I can try to do so.

**Comment by farodin91** [comment]
> @crenshaw-dev I'm not the original author of the PR, @farodin91 is. I'm happy to attempt those things it will help but would need to create a new PR to do so. I'd like to give @farodin91 at least a day or two to respond and if they cannot do it I can try to do so.

Go for it. I'm on vacation and next in this week or next week.

**Comment by asaf-erlich** [comment]
>Go for it. I'm on vacation and next in this week or next week.

I'll give it a try

**Comment by asaf-erlich** [comment]
Here is the rebased PR: https://github.com/argoproj/gitops-engine/pull/565

**Comment by asaf-erlich** [comment]
>Could you also open a PR on argo-cd pointing to the latest commit of your fork/branch? Here's an example of what that change looks like. https://github.com/argoproj/argo-cd/commit/90ce09a75ee7136d08bc11d9c89018e2353d6652
>
>That PR will let us run Argo CD's tests and catch any issues which weren't revealed by the gitops-engine tests.

Looking at the PR and at https://stackoverflow.com/questions/72312460/how-to-replace-a-go-module-with-a-major-version-to-a-fork-master I will need to edit my own fork's go.mod file, at least temporarily, to declare a different name.

**Comment by asaf-erlich** [comment]
My first attempt at the test PR (left it in draft state): https://github.com/argoproj/argo-cd/pull/16967

**Comment by recharte** [comment]
Hi @asaf-erlich and @crenshaw-dev, would you by any chance be able to drive this effort to completion?

I'm in a similar situation as the one described by @asaf-erlich and would greatly benefit from this upgrade. I would even advocate for going a bit further and bumping to 1.29.x, if possible.

**Comment by asaf-erlich** [comment]
Unfortunately I don't have the time to make the required changes.

**Comment by crenshaw-dev** [comment]
Focusing effort on this PR: https://github.com/argoproj/gitops-engine/pull/566

**Comment by blakepettersson** [comment]
Superceded by #566.

**Pull request closed without merging**
